library/src/main/java/com/datatorrent/lib/join/AbstractJoinOperator.java

```diff
@@ -0,0 +1,341 @@
+  private void populateIncludeFields()
+  {
+    includeFields = new String[2][];
+    String[] portFields = includeFieldStr.split(";");
+    for (int i = 0; i < portFields.length; i++) {
+      includeFields[i] = portFields[i].split(",");
+    }
+  }
+
+  /**
+   * Get the tuples from another store based on join constraint and key
+   *
+   * @param tuple  input
+   * @param isLeft whether the given tuple is from first port or not
+   */
+  private void join(TimeEvent tuple, Boolean isLeft)
+  {
+    // Get the valid tuples from the store based on key
+    // If the tuple is null means the join type is outer and return unmatched tuples from store.
+    Object value;
+    if (isLeft) {
+      if (tuple != null) {
+        value = store[1].getValidTuples(tuple);
+      } else {
+        value = store[1].getUnMatchedTuples();
+      }
+    } else {
+      if (tuple != null) {
+        value = store[0].getValidTuples(tuple);
+      } else {
+        value = store[0].getUnMatchedTuples();
+      }
+    }
+    // Join the input tuple with the joined tuples
+    if (value != null) {
+      ArrayList<TimeEvent> joinedValues = (ArrayList<TimeEvent>)value;
+      List<T> result = new ArrayList<>();
+      for (TimeEvent joinedValue : joinedValues) {
+        T output = createOutputTuple();
+        Object tupleValue = null;
+        if (tuple != null) {
+          tupleValue = tuple.getValue();
+        }
+        copyValue(output, tupleValue, isLeft);
+        copyValue(output, joinedValue.getValue(), !isLeft);
+        result.add(output);
+        joinedValue.setMatch(true);
+      }
+      if (tuple != null) {
+        tuple.setMatch(true);
+      }
+      if (result.size() != 0) {
+        outputPort.emit(result);
+      }
+    }
+  }
+
+  // Emit the unmatched tuples, if the strategy is outer join
+  @Override
+  public void endWindow()
+  {
+    if (strategy.equals(JoinStrategy.LEFT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN)) {
+      join(null, false);
+    }
+    if (strategy.equals(JoinStrategy.RIGHT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN)) {
+      join(null, true);
+    }
+    store[0].endWindow();
+    store[1].endWindow();
+  }
+
+  @Override
+  public void checkpointed(long windowId)
+  {
+    store[0].checkpointed(windowId);
+    store[1].checkpointed(windowId);
+  }
+
+  @Override
+  public void committed(long windowId)
+  {
+    store[0].committed(windowId);
+    store[1].committed(windowId);
+  }
+
+  /**
+   * Convert the given tuple to event
+   *
+   * @param tuple Given tuple to convert into event
+   * @return event
+   */
+  protected TimeEvent createEvent(Object tuple)
+  {
+    int idx = 0;
+    if (!isLeft) {
+      idx = 1;
+    }
+    if (timeFields != null) {
+      return new TimeEventImpl(getKeyValue(keys[idx], tuple), (Long)getTime(timeFields[idx], tuple), tuple);
+    } else {
+      return new TimeEventImpl(getKeyValue(keys[idx], tuple), Calendar.getInstance().getTimeInMillis(), tuple);
```

Comment:
If I have to do join on multiple fields?

If you are not supporting join on multiple fields can please you mention it clearly in the Java doc that List contains only two items... Can you also size check in populateKeyFields and populateTimeFields?